### ISSUE TITLE
Pester : Assert-MockCalled not returning expected value 

### ISSUE THREAD

**Issue by maxcli** [pester/Pester#598]
Hello,
I am mocking  a function  like this : 
Context "aaa"
       mock functionxx {"some string"}
      $result =(&$sut ......
      it "**Verify  functionxx mock was called one time"  {
                Assert-MockCalled    functionxx -Exactly 1
       }

The test fails   : expected  1 times exactly but was called 7 times <-- this number seems to change randomly. How do i troubleshoot this ? in  $sut , i have a breakpoint where I call the fuctionxx - just one time.
Thanks , Peter

**Comment by maxcli** [comment]
Hello,
i observe that Pester test (Assert-MockCalled functionxx -Exactly 1) passes if i run from Powershell ISE. But fails  when i run from VS Code.  Any ideas why ?

**Comment by dlwyatt** [comment]
No idea, sorry.  :(  I'd need to see your code to be able to debug what's happening.

**Comment by EvilTerrier** [comment]
Hi,

I'm a little new to Pester (and GitHub ;o) so I can't completely rule out an error on my part, but I seem to be seeing similar issue to that raised in this issue/item/defect.

I'm using Visual Studio 2015 (with Update 3), the PowerShell Tools for VS2015 (v3.0.999) and requested Pester-related, version information is as follows:

- Pester version  : 4.0.3 
- (Path): C:\Users\paul\Documents\WindowsPowerShell\Modules\Pester\4.0.3\Pester.psm1
- PowerShell version : 5.1.14393.1066
- OS version : Microsoft Windows NT 10.0.14393.0


I've a number of simple test cases around a few simple functions in which I am mocking some functions and trying to ensure they are being called (i.e. using "Assert-MockCalled").

I'd been trying to debug some issues with failing tests I was having (it's taken most of the day) but having investigated further, it seems some of the tests I had written **_WERE_** written as I would expect them to work correctly, but were failing as part of the execution within the VS2015 PowerShell test runner tools/components.

The prime case in point was the "It" labelled...
- "SHOULD call Get-MicrosoftSqlServerDacDllDefaultFilePaths function EXACTLY once" 
...in...
- "Get-MicrosoftSqlServerDacDllFilePathFromDefaultFilePaths.tests.ps1"

```
		It "SHOULD call Get-MicrosoftSqlServerDacDllDefaultFilePaths function EXACTLY once" {

			$fileExistsFilePath = "TestDrive:\FileExists.dll"
			Set-Content -Path $fileExistsFilePath -Value "TestContent"
			Mock Get-MicrosoftSqlServerDacDllDefaultFilePaths {"TestDrive:\NoDir1\NoFile1.dll",$fileExistsFilePath,"TestDrive:\NoDir3\NoFile3.dll"} 
			
			Get-MicrosoftSqlServerDacDllFilePathFromDefaultFilePaths
		
			Assert-MockCalled Get-MicrosoftSqlServerDacDllDefaultFilePaths -Times 1 -Exactly -Scope It
			
		}
```

When run in ISE (after a lot of failed debugging in VS2015), I see nothing but passes:

![pester_bug598_iseoutput](https://cloud.githubusercontent.com/assets/28236991/25568945/ef969e06-2e04-11e7-98a8-e0def46389af.png)

...though VS2015 shows...

![pester_bug598_vs2015output](https://cloud.githubusercontent.com/assets/28236991/25568944/ef9436fc-2e04-11e7-92bc-846c4d4a2f8c.png)

...as if the call to the Mock function has not been registered (in the above "It", and another). I'd also had a similar problem where my mock function wasn't returning any values when it was in ISE - i.e. `returning` 3 string values/objects returned in the tests run in ISE, but not in VS2015. 

I appreciate that this may be a "PowerShell Tools for VS2015" issue problem and not a Pester one (it may need somebody here to confirm), but I'm a little unclear why some of the mocking functionality seems to be being ignored (could it be the context of execution or something relating to scope I'm using or the test runner is using, or even some local settings on my machine). 

I've attached a cut-down version of the VS2015 solution I was using that contains two projects (one containing module and functions, the other containing the test scripts per function):

[Pester_Bug598_PJW.Library.PesterExample.zip](https://github.com/pester/Pester/files/967088/Pester_Bug598_PJW.Library.PesterExample.zip)

I hope somebody can provide a bit of feedback on this. Please let me know if you've any questions.

Cheers,

Paul

**Comment by EvilTerrier** [comment]
Just as an update to this thread/issue as I've also posted quite a bit in this issue/thread  (https://github.com/adamdriscoll/poshtools/issues/512)...

It seems as though there is a difference in functionality of "Invoke-Pester" mocking functions dependent upon how the function(s) being tested (and/or mocked) are imported/sourced.

Methods of sourcing functions used/tried: 
 - 1) (dot) source the function under test and the functions to be mocked as part of test script
 - 2) import a module ("psd1" file), where the "psd1" file uses the "psm1" file to (dot) source all the functions within the module

Methods of executing tests (trying to ensure no cached modules/functions etc. each time):
 - A) VS2015 Test Explorer  - POSHTools (believe this uses "Invoke-Pester" underneath)
 - B) VS2015 F5 (running test script in VS2015 window)
 - C) PS ISE - Running test script directly
 - D) PS ISE - Running "Invoke-Pester", referencing test script

In the example tests tried/investigated (the test project for this is attached in my original post), the tests that I would all expect to pass, DO pass in each of A, B, C and D when using the original method of sourcing the functions (method i).

HOWEVER, when using...
- method (ii) to source the functions (i.e. import the module that contains the functions) and;
- using either "A" or "D" execution methods (both use "Invoke-Pester")the mock functions
...the mock functions created in the test script _seem_ to be created as per the pester script/definition but execution of these mocks seems to fail. 

For example, "Assert-MockCalled", seems to be 0 (and fails the test assertion) when using the sourcing method described in (i), but when using the sourcing method described in (ii), this returns 1, as expected (and passes the test assertion).


Other points that may be relevant:
 - I did have a little try to see if using the ModuleName argument was relevant here. Trying to use this on creation and assertion of the mock was giving me ""path" was null" type errors so didn't get far with this
 - I've all the functions in their own ".ps1" files and all the tests for those functions in their own ".tests.ps1" files. The "psd1" file references the "psm1" module and this module imports (dot-sources) all the functions as the module script/body. I'm importing the module at the start of each test script via a function (details on https://github.com/adamdriscoll/poshtools/issues/512)

As mentioned, more information here (https://github.com/adamdriscoll/poshtools/issues/512) and a zip with the VS2015 solution I was using is in my original post/cooment.

Please let me know if you've any questions (and/or if you've any thoughts/comments/workarounds etc.).

Cheers,

Paul

**Comment by nohwnd** [comment]
@SphenicPaul Thanks for reporting back, if you still experiencing this issue please feel free to re-open it and we can come up with a simple repro together. I also recently merged #1169 that allows you to print in which scope which block executes because Pester jump between scopes a lot and so it is hard to see what variables  should and should not be available to the test code.
